src/generate_cropped_dataset.py

```diff
@@ -0,0 +1,335 @@
+        self.current_gt_img = pil_loader(self.img_paths[img_index][1])
+
+        # Update pointer to current image
+        self.current_img_index = img_index
+
+    def get_crops(self, img, gt, coordinates, crop_function):
+        img, gt = crop_function(img, gt, coordinates)
+        return img, gt
+
+    def _get_img_size_and_crop_numbers(self):
+        # TODO documentation
+        img_names_sizes = []  # list of tuples -> (gt_img_name, img_size (H, W))
+        num_horiz_crops = []
+        num_vert_crops = []
+
+        for img_path, gt_path in self.img_paths:
+            data_img = pil_loader(img_path)
+            gt_img = pil_loader(gt_path)
+            # Ensure that data and gt image are of the same size
+            assert gt_img.size == data_img.size
+            img_names_sizes.append((os.path.basename(gt_path), data_img.size[::-1]))
+            step_size = self.crop_size * self.overlap
+            num_horiz_crops.append(math.ceil((data_img.size[1] - self.crop_size) / step_size + 1))
+            num_vert_crops.append(math.ceil((data_img.size[0] - self.crop_size) / step_size + 1))
+
+        return img_names_sizes, num_horiz_crops, num_vert_crops
+
+    def _get_crop_list(self):
+        return [self._convert_crop_id_to_coordinates(img_index, hcrop_index, vcrop_index) for img_index in
+                range(self.num_imgs_in_set) for hcrop_index, vcrop_index in
+                itertools.product(range(self.num_horiz_crops[img_index]),
+                                  range(self.num_vert_crops[img_index]))]
+
+    def _convert_crop_id_to_coordinates(self, img_index, hcrop_index, vcrop_index):
+        # X coordinate
+        if hcrop_index == self.num_horiz_crops[img_index] - 1:
+            # We are at the end of a line
+            x_position = self.img_names_sizes[img_index][1][0] - self.crop_size
+        else:
+            x_position = int(self.crop_size / (1 / self.overlap)) * hcrop_index
+            assert x_position < self.img_names_sizes[img_index][1][0] - self.crop_size
+
+        # Y coordinate
+        if vcrop_index == self.num_vert_crops[img_index] - 1:
+            # We are at the bottom end
+            y_position = self.img_names_sizes[img_index][1][1] - self.crop_size
+        else:
+            y_position = int(self.crop_size / (1 / self.overlap)) * vcrop_index
+            assert y_position < self.img_names_sizes[img_index][1][1] - self.crop_size
+
+        return img_index, x_position, y_position
+
+
+if __name__ == '__main__':
+    dataset_generator = CroppedDatasetGenerator(
+        input_path=Path('/dataset/DIVA-HisDB/segmentation/CB55-10-segmentation'),
+        output_path=Path('/data/usl_experiments/semantic_segmentation/datasets_cropped/CB55-10-segmentation'),
+        crop_size_train=300,
+        crop_size_val=300,
+        crop_size_test=256,
+        overlap=0.5,
+        leading_zeros_length=4,
+        override_existing=True)
+    dataset_generator.write_crops()
```

Comment:
Perhaps we could create a cli out of it (like we are doing in the other scripts)

Reply:
Good idea.